README.md
## FayeCom
This is a nodejs app that allows for passing messages to clients using efficient websockets on an as-needed basis.  

## Install

# Install nodejs (compile from source)

# Install npm (compile from source)

# Install fayecom and its dependencies globally
    $  npm install -g

## Configuration

Configuration values are just at the top of app.js.
You can specify the port to run on, the name of the environment variable that contains the authentication token, ...

You'll want to create some ssl keys, this explains the procedure:  http://stackoverflow.com/questions/12871565/how-to-create-pem-files-for-https-web-server
    $  cd test/fixtures/keys
    $  openssl req -newkey rsa:2048 -new -nodes -keyout key.pem -out csr.pem
    $  openssl req -newkey rsa:2048 -new -nodes -x509 -days 3650 -keyout key.pem -out cert.pem

The development keys are locked into this repo, and went public, so remake teh keys in production.  


### server only publishing
This app is designed to allow only the server to publish messages to the subscribers.  
This rule is enforced with a secret authentication token named 

## Usage

Start the app.  

    `$  fayecom`

Browser clients can now connect.  
This system is designed to interact with MANY browser clients, and usually ~1 ruby server client (which signals to all the clients).  
Requests to /version will be restponded to with the version of fayecom.  


The best way to start it is with forever.  Just plop in some code to boot forever on @reboot in `crontab`
    
    $  /usr/local/bin/forever start /usr/local/bin/fayecom



    


## Testing

Boot the app from the dev dir by `$  lib/fayecom.js`

The below snippet was super helpful for testing if the server was online.

    // Browser test Client Snippet, logs 3 if successful connection
    var channel = "/save_kittens/data/fresh_data";
    var url = location.protocol + "//" + document.domain + ":8000/faye"; 
    var fayeClient = new Faye.Client(url);
    var subscription = fayeClient.subscribe(channel, function(message) {
      // message: {"signatureCount":54,"topThreeStates":["CA"]}
      alert(message);
    });

    setTimeout(function(){
      console.log(fayeClient._state);
    }, 500);



## Node Cliffnotes for Rails Hackers

  - nodejs is unique in that it is based around the concept of non-blocking code, not just that it's js for server side
    > a node app runs as a signle process, a single thread, node has no concept of threads
    > Node apps are parsed, and all your code does is register events.  
    > after your app evaluates, your app enters it's event loop where events are handled accordingly as they come into the event que (single thread)
    >>> That's how a single threaded node app can listen on multiple ports, those listen calls are non-blocking, they register events

  - package.json lists dependencies, project metadata, the 'entrypoint' or main js file, and project version

  - node_modules/ is where dependencies are installed to when you type `npm install`, don't version track it

  - To debug on breakpoints, start with the parameter debug, `node debug app.js`
    > `cont` will continue to the next `debugger;` statement in your code
    > `repl` will allow you to interact with the environment at your current position

  :: Oragnization of Code ::
    - Modules - 
      > require("yourfile.js") will return a module composed of methods.
        You assign methods to a module via `exports.myMethod= function(){return "hello module"}`
        var o = require("yourfile.js").myMethod();
        // => "hello module"
        ref: http://kevgriffin.com/node-js-using-require-to-load-your-own-files/

  - Misc dev hints
    - https://www.airpair.com/node.js/posts/top-10-mistakes-node-developers-make


  - Using sysvinit (default debian)
    > https://www.npmjs.com/package/initd-forever
  - Using upstart (default ubuntu/ linux mint)
    > https://www.npmjs.com/package/upstarter
  - Using forever (cli)
    > http://blog.nodejitsu.com/keep-a-nodejs-server-up-with-forever/






changelog.md
# 0.0.x

  - the server will host files of http or https when an https server is
    configured, but it won't do websockets correctly over http when https
    is selected...

currently:

  - implement a configuration file

  - implement daemon -d option



finished:

  - CORE:  restructured project layout, now distributed as binary package
  - API: added /version for looking up fayecom version

# 0.0.1
  - implemented versioning
  - Uses system of serving sockets on 8000 and ssl mode 8001
  - Authenticates publishers



test_client.js
var http = require('http'),
    faye = require('faye');

var client = new faye.Client('http://localhost:8000/faye');

client.subscribe('h1');

//client.publish('h1', "hello");


app.js
var port = 8000,
    authTokenEnvName = "SAVE_KITTENS_FAYE_TOKEN";

var authToken = process.env[authTokenEnvName];

var https = require('https'),
    http = require('http'),
    url=require('url'),
    fs = require('fs'),
    faye = require('faye'),
    pjson = require('./package.json');

// This line is from the Node.js HTTPS documentation.
var options = {
  key: fs.readFileSync('test/fixtures/keys/key.pem'),
  cert: fs.readFileSync('test/fixtures/keys/cert.pem')
};

var tlsServer = https.createServer(options, function(request, response){
      var pathname = url.parse(request.url, true).pathname;
      if (pathname == "/version") response.end(pjson.version);
    }),
    server = http.createServer(function(request, response){
      var pathname = url.parse(request.url, true).pathname;
      if (pathname == "/version") response.end(pjson.version);
    }),
    bayeux = new faye.NodeAdapter({mount: '/faye', timeout: 45});

// Echo messages the server receives
bayeux.on('publish', function(clientId, channel, data) {
  console.log(xinspect(data));
});

// Ensures only the rails server can publish by using secret token
var serverOnlyWriteAccess = {
  incoming: function(message, callback) {

    // Let non-publishing messages through
    if (message.channel !== '/save_kittens/data/fresh_data')
      return callback(message);

    // Get subscribed channel and auth token
    var subscription = message.subscription,
        msgToken     = message.ext && message.ext.authToken;
    
    // Add an error if the tokens don't match
    if (authToken !== msgToken)
      message.error = 'Invalid publication auth token';

    // Call the server back now we're done
    callback(message);
  }

};

console.log("FayeCom " + pjson.version + "\nstarting server");

bayeux.addExtension(serverOnlyWriteAccess);

bayeux.attach(server);
server.listen(port);

bayeux.attach(tlsServer);
tlsServer.listen(port+1);



function xinspect(o,i){
    if(typeof i=='undefined')i='';
    if(i.length>50)return '[MAX ITERATIONS]';
    var r=[];
    for(var p in o){
        var t=typeof o[p];
        r.push(i+'"'+p+'" ('+t+') => '+(t=='object' ? 'object:'+xinspect(o[p],i+'  ') : o[p]+''));
    }
    return r.join(i+'\n');
}



lib/fayecom.js
var port = 8000,
    channels = [
      { name: '/save_kittens/data/fresh_data', 
        authenticate: { subscription: false, publication: true }
      }
    ],
    authTokenEnvName = "SAVE_KITTENS_FAYE_TOKEN";

var authToken = process.env[authTokenEnvName];

var https = require('https'),
    http = require('http'),
    url=require('url'),
    fs = require('fs'),
    path = require('path'),
    faye = require('faye'),
    IsThere = require("is-there"),
    pjson = require('./../package.json');

var appRoot = path.resolve(__dirname + "/..");

if (!IsThere(appRoot  + '/test/fixtures/keys/key.pem')){
  console.log("Could not find the .pem key!");
  console.log("Not in " + appRoot);
  process.exit(1);
}

var tlsOptions = {
  key: fs.readFileSync(appRoot + '/test/fixtures/keys/key.pem'),
  cert: fs.readFileSync(appRoot + '/test/fixtures/keys/cert.pem')
};

var tlsServer = https.createServer(tlsOptions, function(request, response){
      var pathname = url.parse(request.url, true).pathname;
      if (pathname == "/version") response.end(pjson.version);
    }),
    server = http.createServer(function(request, response){
      var pathname = url.parse(request.url, true).pathname;
      if (pathname == "/version") response.end(pjson.version);
    }),
    bayeux = new faye.NodeAdapter({mount: '/faye', timeout: 45});

// Echo messages the server receives
bayeux.on('publish', function(clientId, channel, data) {
  console.log(xinspect(data));
});

// Ensures only the rails server can publish by using secret token
var serverOnlyWriteAccess = {
  incoming: function(message, callback) {

    // Let non-publishing messages through
    if (message.channel !== '/save_kittens/data/fresh_data')
      return callback(message);

    // Get subscribed channel and auth token
    var subscription = message.subscription,
        msgToken     = message.ext && message.ext.authToken;
    
    // Add an error if the tokens don't match
    if (authToken !== msgToken)
      message.error = 'Invalid publication auth token';

    // Call the server back now we're done
    callback(message);
  }

};

console.log("FayeCom " + pjson.version + "\nstarting server");

bayeux.addExtension(serverOnlyWriteAccess);

bayeux.attach(server);
server.listen(port);

bayeux.attach(tlsServer);
tlsServer.listen(port+1);



function xinspect(o,i){
    if(typeof i=='undefined')i='';
    if(i.length>50)return '[MAX ITERATIONS]';
    var r=[];
    for(var p in o){
        var t=typeof o[p];
        r.push(i+'"'+p+'" ('+t+') => '+(t=='object' ? 'object:'+xinspect(o[p],i+'  ') : o[p]+''));
    }
    return r.join(i+'\n');
}

